R/pmcmc_control.R

```diff
@@ -0,0 +1,142 @@
+##' Control for the pmcmc. This function constructs a list of options
+##' and does some basic validation to ensure that the options will
+##' work well together. Do not manually change the values in this
+##' object. Do not refer to any argument except `n_steps` by position
+##' as the order of the arguments may change in future.
+##'
+##' @title Control for the pmcmc
+##'
+##' @param n_steps Number of MCMC steps to run. This is the only
+##'   required argument.
+##'
+##' @param n_chains Optional integer, indicating the number of chains
+##'   to run. If more than one then we run a series of chains and
+##'   merge them with [pmcmc_combine()]. Chains are run in series,
+##'   with the same filter if `n_workers` is 1, or run in parallel
+##'   otherwise.
+##'
+##' @param n_workers Number of "worker" processes to use to run chains
+##'   in parallel. This must be greater than `n_chains` and is
+##'   recommended to be a divisor of `n_chains`. If `n_workers` is 1,
+##'   then chains are run in series (i.e., one chain after the other).
+##'
+##' @param n_steps_each If using workers (i.e., `n_workers > 1`), the
+##'   number of steps to run in each "chunk" on each worker before
+##'   reporting back to the main process. Increasing this will make
+##'   progress reporting less frequent and reduce some communication
+##'   overhead (though the overhead is likely to be trivial in any
+##'   real application). Decreasing this will give more frequent
+##'   process reporting and if `n_threads_total` is given will allow
+##'   for more rapid re-allocation of unused cores once chains start
+##'   finishing. The default, if not given and if `n_workers > 1` is
+##'   to use `n_steps` which gives *no* within-chain feedback and no
+##'   option to reallocate cores.
+##'
```

Comment:
Needs to scale with the number of steps and is really quite important to set. If you set too high you get no work stealing, too low and too much overhead

Reply:
What about setting to 0.1 * n_steps as the default?

Reply:
I am sure for sircovid this will get tuned to some extent, but I can imagine for many uses this being difficult to set correctly. Do you do a long run with various values to optimise it? I think many things will just be run once, or there will be other parameters that need to be tuned first.
Based on your comment having at 1 or n_chains is going to be suboptimal, so a default somewhere in the middle makes more sense to me, and then those that need to can try and refine the value

Reply:
10% seems as good as any other punt, I can set that. It's something that I think would be most usefully set in units of *seconds* but that's not really possible without really complicating the mcmc code itself (like, report back in every minute).